CHANGES.md

```diff
@@ -4,6 +4,7 @@
 
 ### Features
 
+* A new command, `augur merge`, now allows for generalized merging of two or more metadata tables. [#1563][] (@tsibley)
```

Comment:
What did we say about supporting sequence merge? I remember talking about this but forgot the conclusion.

Reply:
We took it off the table for `augur merge`, at least initially. If we add it, it'd use the `augur filter` machinery to filter sequences by strain based on the output metadata. Doing so manually for now would roughly look something like:

```bash
augur merge --metadata a=a.tsv b=b.csv --output-metadata c.tsv
cat a.fasta b.fasta > all.fasta
augur filter --metadata c.tsv --sequences all.fasta --output-sequences c.fasta
```

Reply:
Unfortunately, looking at the code, I think `augur filter` doesn't support `--sequences` from a non-seekable stream.

Reply:
If the implementation is as simple as the workaround, I think we should add built-in support upon initial release.

Most (or all?) workflows require a pair of metadata+sequences files. That means if we release a metadata-only `augur merge`, we would end up applying the manual workaround above everywhere we start using it, and it'd be an extra thing to go back and update if we end up adding built-in support down the road.

Reply:
Well, the implementation in `augur merge` gets trickier because we need to consider that someone may reasonably give `--output-metadata -` and we won't have an output file to re-read as in the workaround. It's all doable, but definitely extra work that's not quite as straightforward.

Reply:
> because we need to consider that someone may reasonably give `--output-metadata -`

I'm not sure that this needs to be considered. Commands that use `augur.io.read_metadata` don't support metadata input from a non-seekable stream:

```
$ augur merge \
    --metadata x=x.tsv y=y.tsv \
    --output-metadata - \
| augur filter \
    --metadata - \
    --subsample-max-sequences 1 \
    --output-metadata filtered.tsv
Reading 'x' metadata from 'x.tsv'…
Reading 'y' metadata from 'y.tsv'…
Merging metadata and writing to '-'…
ERROR: -: Expected a header row but it is empty.
```

Unless I'm missing some other use case, `--output-metadata -` would only be useful for debugging purposes and not for a typical workflow.

I see `augur merge` as similar to `augur filter`, which supports metadata and/or sequences and can do `--output-metadata -` if that's the only output that's requested.

Reply:
ISTM there are plenty of reasons you might use `--output-metadata -` to pipe to a program that does accept non-seekable streams as input, for example, the `augur curate` commands or many non-Augur commands.

Reply:
Fair enough. I think in order to make a decision here, we should determine `augur merge`'s expected placement in a workflow. If it's before `augur curate`, the current implementation seems perfect. If it's before `augur filter`, it would be good to support sequences.

Reply:
Agree with Tom that being able to input from/output to stdin/out is useful in many ways and should be supported by default unless there are good reasons not to.

Reply:
Merging sequences is critical being able to accomplish the goal of spiking in user-data with reference-data. This would generically look like a rule between [download/decompress reference sequences from S3](https://github.com/nextstrain/measles/blob/main/phylogenetic/rules/prepare_sequences.smk#L20) and [filter/subsample](https://github.com/nextstrain/measles/blob/main/phylogenetic/rules/prepare_sequences.smk#L34).

This rule needs to combine user-supplied sequence FASTA with reference FASTA and needs to combine user-supplied metadata TSV with reference TSV.

I think it's an acceptable thing to initially ship `augur merge` with only support for `--metadata` + `--output-metadata` and to add `--sequences` + `--output-sequences` shortly later.

I _think_ that `cat reference.fasta user.fasta > all.fasta` will largely suffice as a work around here, as although `all.fasta` may contain duplicates they should be stripped in the later `augur filter` call.

The use case it doesn't tackle is where the user has updated a particular entry in `user.fasta` and would like to overwrite entry in `reference.fasta`. Ie do the same sort of ["keep right"](https://github.com/nextstrain/augur/pull/1563/files#diff-d032c8de9ed6a70822d4a862a85f9e6aaff140bfd8e377d2d09a7a77fca73a07R14) behavior for sequences as for columns. I imagine this particular use case isn't super common and so starting with `cat reference.fasta user.fasta > all.fasta` should be okay.

Reply:
Ah! I just tested and I was wrong here. `augur filter` doesn't drop duplicates. I just ran zika with duplicate sequences and the final resulting tree possessed duplicates.

So I don't think we could add `augur merge` metadata + `cat` sequences  between download and filter and expect to work generically to spike-in user data. We could still plan to initially ship `augur merge` for metadata only, but we'll need support for sequences to follow pretty shortly for intended use case.

Reply:
I made a note of the surprising (at least to me) behavior of allowing duplicate sequences in https://github.com/nextstrain/augur/issues/810#issuecomment-2285166344.

Reply:
I wondered how ncov handled duplicate sequences, because ncov explicitly supports the merging/combining of multiple inputs already and its `scripts/combine_metadata.py` is in large part the motivation for `augur merge` to exist more generally.

ncov [deduplicates sequences via `scripts/sanitize_sequences.py`](https://github.com/nextstrain/ncov/blob/0769ac0429df8456ce70be2f74dc885d7b7fab12/scripts/sanitize_sequences.py#L42-L85) using [a rule that's only called when multiple inputs are involved](https://github.com/nextstrain/ncov/blob/0769ac0429df8456ce70be2f74dc885d7b7fab12/workflow/snakemake_rules/main_workflow.smk#L214-L238) (paralleling how `scripts/combine_metadata.py` is used).

Reply:
Given the details of sequence merging that we haven't considered (because we'd taken it off the table in the original `augur merge` discussions), I'd prefer to ship this PR without sequencing merging and add it later after further discussion and implementation ideas.

Reply:
Definitely support this decision.

Reply:
https://github.com/nextstrain/augur/issues/1579